File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
      auto pair = helperDistance(root);
      std::cout << pair.first << pair.second << std::endl;
      return (pair.first > pair.second) ? pair.first : pair.second;
    }

    std::pair<int,int> helperDistance(TreeNode* node) {
      std::pair left = std::pair(0,0);
      std::pair right = std::pair(0,0);

      if (node->left != nullptr) {
        left = helperDistance(node->left);
        left.first++;
      }

      if (node->right != nullptr) {
        right = helperDistance(node->right);
        right.first++;
      }

      int maxDepth = (left.first > right.first) ? left.first : right.first;
      int maxWidth = (left.second > right.second) ? left.second : right.second;
      maxWidth = (maxWidth > left.first + right.first) ? maxWidth : left.first + right.first;


      return std::pair(
        maxDepth,
        maxWidth
      );
    }
};

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.java
class Solution {
    public int maxDepth(TreeNode root) {
      return maxDepthHelper(root, 0);

    }

    private int maxDepthHelper(TreeNode node, int depth) {
      if (node == null) return depth;

      int newDepth = depth + 1;
      int left = maxDepthHelper(node.left, newDepth);
      int right = maxDepthHelper(node.right, newDepth);

      return (right > left) ? right : left;
    }
}
